### PR TITLE
GPU local texture data copying

### PR DESCRIPTION
**Pull request by binary1248**
In many applications, it is common for the user to stitch smaller textures into bigger ones in order to minimize the amount of OpenGL state changes necessary when drawing. This is commonly known as atlasing. Even `sf::Font` makes use of this when loading its glyphs.

The problem with atlasing in SFML is that each time a new texture has to be inserted into the atlas, users would have to copy the texture data back to RAM, resize/update it and transfer it back on to the GPU. This costs a lot of time. With this PR, the user (given the proper hardware support) will be able to copy the existing texture to a temporary texture on the GPU, resize the original texture to its new size and copy the old texture data over. This does not involve any copying of data between the GPU and CPU.

``` cpp
#include <SFML/Graphics.hpp>

int main()
{
    sf::Font font;
    if (!font.loadFromFile("resources/sansation.ttf"))
        return -1;

    for (unsigned int i = 32; i < 127; i++)
        font.getGlyph(i, 1000, false, 0);
}
```

Testing with this code yields 2 orders of magnitude performance increase. On my system, the time it takes for the font to load the codepoints decreases from 20 seconds down to under 1 second.

This is only indirectly testing the texture copying and is a contrived example, since it is meant to be a minimal demonstration. If anybody else has texture copy benchmarks, they can provide the data here as well.

In addition to this feature, I went ahead and optimized `sf::Font::cleanup()` to deallocate its pixel buffer storage. This way, if the user loads glyphs using a huge character size and later reuses the same `sf::Font` object for another font at smaller character sizes, they won't be stuck with a `std::vector` using up more memory than necessary.

### CODE REVIEW

File: src/SFML/Graphics/Texture.cpp [hunk 1]
```diff
@@ -441,6 +454,94 @@ void Texture::update(const Uint8* pixels, unsigned int width, unsigned int heigh
...
+        glCheck(GLEXT_glBindFramebuffer(GLEXT_GL_READ_FRAMEBUFFER, readFramebuffer));
+        glCheck(GLEXT_glBindFramebuffer(GLEXT_GL_DRAW_FRAMEBUFFER, drawFramebuffer));
+
+        // Delete the framebuffer
+        glCheck(GLEXT_glDeleteFramebuffers(1, &frameBuffer));
+
+        return;
+    }
+
+#endif // SFML_OPENGL_ES
```
Comment (mantognini) [inline src/SFML/Graphics/Texture.cpp]: Isn't this supposed to be an `#else` to handle the fallback just below?

**Comment by mantognini** [comment]
Interesting, I'll test this code in a bit to see how it affect perfs on my machine! :-)

File: src/SFML/Graphics/Texture.cpp [hunk 2]
```diff
@@ -441,6 +454,94 @@ void Texture::update(const Uint8* pixels, unsigned int width, unsigned int heigh
...
+        glCheck(GLEXT_glFramebufferTexture2D(GLEXT_GL_READ_FRAMEBUFFER, GLEXT_GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, texture.m_texture, 0));
+
+        // Link the destination texture to the frame buffer
+        glCheck(GLEXT_glFramebufferTexture2D(GLEXT_GL_DRAW_FRAMEBUFFER, GLEXT_GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, m_texture, 0));
+
+        // A final check, just to be sure...
+        GLenum status;
+        glCheck(status = GLEXT_glCheckFramebufferStatus(GLEXT_GL_FRAMEBUFFER));
+
+        if (status != GLEXT_GL_FRAMEBUFFER_COMPLETE)
```
Comment (mantognini) [inline src/SFML/Graphics/Texture.cpp]: might be more concise to refactor as follows:

```
if (status == GLEXT_GL_FRAMEBUFFER_COMPLETE)
    GLEXT_glBlitFramebuffer(...)
else
    err() << ...

cleanup / return

File: src/SFML/Graphics/Texture.cpp (hunk 1 above)
Comment (mantognini) [inline src/SFML/Graphics/Texture.cpp]: ho... I just noticed the `return` above and it's implication with the `if (GLEXT_framebuffer_object && GLEXT_framebuffer_blit)`... a bit spaghetti. 😜

File: src/SFML/Graphics/Texture.cpp [hunk 3]
```diff
@@ -441,6 +454,94 @@ void Texture::update(const Uint8* pixels, unsigned int width, unsigned int heigh
...
 ////////////////////////////////////////////////////////////
+void Texture::update(const Texture& texture)
+{
+    // Update the whole texture
+    update(texture, 0, 0);
+}
+
+
+////////////////////////////////////////////////////////////
+void Texture::update(const Texture& texture, unsigned int x, unsigned int y)
```
Comment (MarioLiebisch) [inline src/SFML/Graphics/Texture.cpp]: Wouldn't it be better to just take the offset by accepting `sf::Vector2u` instead? You'll always have both params anyway. You could also just pass `copy.getSize()` in line 112.

**Comment by mantognini** [comment]
Best performance of several runs, without this patch: `1.67 real         0.60 user         0.33 sys`
Worst performance of several runs, with this patch: `1.12 real         0.14 user         0.08 sys`

👍

File: src/SFML/Graphics/Texture.cpp (hunk 1 above)
Comment (binary1248) [inline src/SFML/Graphics/Texture.cpp]: I couldn't think of any other way of writing it so that even on non-ES systems it would fall through to the CPU copy if the extension isn't available...

File: src/SFML/Graphics/Texture.cpp (hunk 3 above)
Comment (binary1248) [inline src/SFML/Graphics/Texture.cpp]: I tried to stay consistent with the other update methods and they all take `unsigned int`s instead of `sf::Vector2u`s. I assume there was/is a reason for that, so the same reason will apply here as well.

File: src/SFML/Graphics/Texture.cpp (hunk 2 above)
Comment (binary1248) [inline src/SFML/Graphics/Texture.cpp]: Fixed.

File: src/SFML/Graphics/Texture.cpp (hunk 1 above)
Comment (mantognini) [inline src/SFML/Graphics/Texture.cpp]: I'm not familiar with OpenGL ES enough to know if the following would compile... but it could be more readable.

```cpp
#ifdef SFML_OPENGL_ES
    bool fallback = true;
#else
    ensureGlContext();
    priv::ensureExtensionsInit();
    bool fallback = !(GLEXT_framebuffer_object && GLEXT_framebuffer_blit);
#endif

if (fallback) update(texture.copyToImage(), x, y);
else {
 // the big part here
}
```

File: src/SFML/Graphics/Texture.cpp (hunk 3 above)
Comment (mantognini) [inline src/SFML/Graphics/Texture.cpp]: I think for now the way @binary1248 did it is good, but we should reflect on this for 3.0.

File: src/SFML/Graphics/Texture.cpp (hunk 1 above)
Comment (binary1248) [inline src/SFML/Graphics/Texture.cpp]: If you ask me, this makes the code harder to reason about than the current version.

File: src/SFML/Graphics/Font.cpp [hunk 4]
```diff
@@ -583,11 +583,14 @@ Glyph Font::loadGlyph(Uint32 codePoint, unsigned int characterSize, bool bold, f
         // pollute them with pixels from neighbors
         const unsigned int padding = 1;
 
+        width += 2 * padding;
+        height += 2 * padding;
```
Comment (LaurentGomila) [inline src/SFML/Graphics/Font.cpp]: (sorry I was away the whole week, I try to catch up)

What's the reason for adding 2 * padding to width and height here, and then constantly subtracting it everywhere width and height are used?

Comment (binary1248) [inline src/SFML/Graphics/Font.cpp]: The block of memory we work with now includes the border padding (which it didn't before). Instead of initializing the texture to the padding colour when we create it (which might waste time when the texture gets really big), we do it every time a glyph would actually need the padding around it and upload it then.

**Comment by eXpl0it3r** [comment]
This PR needs a rebase.

File: src/SFML/Graphics/Font.cpp [hunk 5]
```diff
@@ -603,43 +606,43 @@ Glyph Font::loadGlyph(Uint32 codePoint, unsigned int characterSize, bool bold, f
...
-        unsigned int x = glyph.textureRect.left;
-        unsigned int y = glyph.textureRect.top;
-        unsigned int w = glyph.textureRect.width;
-        unsigned int h = glyph.textureRect.height;
-        page.texture.update(&m_pixelBuffer[0], w, h, x, y);
+        unsigned int x = glyph.textureRect.left - padding;
+        unsigned int y = glyph.textureRect.top - padding;
+        unsigned int w = glyph.textureRect.width + 2 * padding;
+        unsigned int h = glyph.textureRect.height + 2 * padding;
+        page.texture.update(reinterpret_cast<const Uint8*>(&m_pixelBuffer[0]), w, h, x, y);
```
Comment (LaurentGomila) [inline src/SFML/Graphics/Font.cpp]: This looks unsafe, since it depends on the processor endianness. On little endian machines you'll end up with RGBA byte ordering and on big endian you'll get ABGR.

Why did you change the pixel buffer from uint8 to uint32?

Comment (binary1248) [inline src/SFML/Graphics/Font.cpp]: The idea was that because the unused area of m_pixelBuffer should contain transparent white pixels, it could be done once during initialization instead of every time a glyph is added. This scales better with bigger glyphs and the more glyphs are added.

The only reason I changed it to Uint32 is because that way initialization would only consist of a single line (the assign). Now that I think of it, since a single memset wouldn't be able to be used internally either (it can only assign a single byte value to a range), the assign would loop over each T anyway. I'll probably revert the storage back to Uint8 and loop initialize using a Uint8 array. If the optimizer isn't too dumb, it should end up producing pretty similar results.

File: src/SFML/Graphics/Font.cpp [hunk 6]
```diff
@@ -693,10 +696,9 @@ IntRect Font::findGlyphRect(Page& page, unsigned int width, unsigned int height)
             if ((textureWidth * 2 <= Texture::getMaximumSize()) && (textureHeight * 2 <= Texture::getMaximumSize()))
             {
                 // Make the texture 2 times bigger
-                Image newImage;
-                newImage.create(textureWidth * 2, textureHeight * 2, Color(255, 255, 255, 0));
-                newImage.copy(page.texture.copyToImage(), 0, 0);
-                page.texture.loadFromImage(newImage);
+                Texture oldTexture(page.texture);
+                page.texture.create(textureWidth * 2, textureHeight * 2);
+                page.texture.update(oldTexture);
```
Comment (binary1248) [inline src/SFML/Graphics/Font.cpp]: Yes. :grin:

**Comment by binary1248** [comment]
At the moment this is not compatible with the context changes i.e. it is broken. Will need to rebase/re-implement/re-test once the context stuff is merged.

**Comment by binary1248** [comment]
Rebased onto master, addressed the raised issues and fixed a few bugs encountered during testing.

File: src/SFML/Graphics/Font.cpp [hunk 7]
```diff
@@ -602,44 +605,57 @@ Glyph Font::loadGlyph(Uint32 codePoint, unsigned int characterSize, bool bold, f
...
-        unsigned int x = glyph.textureRect.left;
-        unsigned int y = glyph.textureRect.top;
-        unsigned int w = glyph.textureRect.width;
-        unsigned int h = glyph.textureRect.height;
-        page.texture.update(&m_pixelBuffer[0], w, h, x, y);
+        unsigned int x = glyph.textureRect.left - padding;
+        unsigned int y = glyph.textureRect.top - padding;
+        unsigned int w = glyph.textureRect.width + 2 * padding;
+        unsigned int h = glyph.textureRect.height + 2 * padding;
+        page.texture.update(reinterpret_cast<const Uint8*>(&m_pixelBuffer[0]), w, h, x, y);
```
Comment (LaurentGomila) [inline src/SFML/Graphics/Font.cpp]: Seems like you forgot to remove the `reinterpret_cast`

Comment (binary1248) [inline src/SFML/Graphics/Font.cpp]: Fixed.

**Comment by eXpl0it3r** [comment]
This seems to work nicely, even though I don't fully understand all the changes, I still approved it, since nobody else has bothered to do so far. 😉 

Here's some code I used to test things a bit:
```cpp
#include <SFML/Graphics.hpp>

int main()
{
    const int gameWidth = 800;
    const int gameHeight = 600;

    sf::RenderWindow window(sf::VideoMode(gameWidth, gameHeight), "SFML Test");
    window.setVerticalSyncEnabled(true);

    sf::Texture tex1;
    tex1.loadFromFile("200.png");
    sf::Sprite spr1(tex1);

    sf::Texture tex2;
    tex2.create(200, 200);
    sf::Sprite spr2(tex2);
    spr2.setPosition({300, 0});

    sf::RenderTexture rt;
    rt.create(200, 200);
    rt.clear(sf::Color::Green);
    rt.display();

    while (window.isOpen())
    {
        sf::Event event;
        while (window.pollEvent(event))
        {
            if (event.type == sf::Event::Closed)
            {
                window.close();
                break;
            }

            if (event.type == sf::Event::KeyPressed)
            {
                if (event.key.code == sf::Keyboard::Space)
                    tex1.swap(tex2);
                else if (event.key.code == sf::Keyboard::Return)
                    tex1.update(rt.getTexture());
            }

        }

        window.clear();
        window.draw(spr1);
        window.draw(spr2);
        window.display();
    }
}
```

**Comment by texus** [comment]
Since this commit was merged the following code now crashes/asserts when copying a texture.

``` c++
#include <SFML/Graphics.hpp>

int main()
{
    sf::Texture texture;
    texture.loadFromFile("image.jpg");

    sf::Texture texture2 = texture;
}
```

Output:
```
SFML/src/SFML/Graphics/Texture.cpp:450: void sf::Texture::update(const sf::Texture&, unsigned int, unsigned int): Assertion `y + texture.m_size.x <= m_size.y' failed.
```

**Comment by eXpl0it3r** [comment]
Thanks for letting us know, I'll see to get this fixed asap.

**Comment by LaurentGomila** [comment]
The fix is trivial, it's just a typo (should be `texture.m_size.y` in the assert condition).

**Comment by vittorioromeo** [comment]
@binary1248: what is the reason that `swap` does not simply swap the cache IDs, instead generating two new ones? Are our move operations incorrect in setting the moved-from texture's cache ID to `0`? See #3077 for context.
